File: home-manager/shell/rice/waifufetch.py
from pathlib import Path
from subprocess import run
import argparse
import json
import os
import signal
import sys


# creates the image and returns the path to the image
def create_image(colors=None):
    if not colors:
        colors = json.load(open(Path("~/.cache/wallust/nix.json").expanduser()))
    logo = colors["neofetch"]["logo"]
    c4 = colors["colors"]["color4"]
    c6 = colors["colors"]["color6"]

    img = f"/tmp/waifufetch-{c4}-{c6}.png".replace("#", "")

    run(
        [
            "magick",
            logo,
            # replace color 1
            "-fuzz",
            "10%",
            "-fill",
            c4,
            "-opaque",
            "#5278c3",
            # replace color 2
            "-fuzz",
            "10%",
            "-fill",
            c6,
            "-opaque",
            "#7fbae4",
            img,
        ]
    )

    return img


def waifufetch():
    colors = json.load(open(Path("~/.cache/wallust/nix.json").expanduser()))
    neofetch_config = colors["neofetch"]["conf"]
    img = create_image(colors)

    run(
        [
            "neofetch",
            "--kitty" if os.environ["TERM"] == "xterm-kitty" else "--sixel",
            img,
            "--config",
            neofetch_config,
        ]
    )


def parse_args():
    parser = argparse.ArgumentParser(
        prog="waifufetch",
        description="Neofetch with more waifu",
    )
    parser.add_argument(
        "--image", action="store_true", help="returns path to generated image"
    )

    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()

    if args.image:
        print(create_image())
        exit()

    def sigusr2_handler(sig, frame):
        waifufetch()

    def sigint_handler(sig, frame):
        sys.exit()

    # initial run
    waifufetch()

    # reload waifufetch on SIGUSR2
    signal.signal(signal.SIGUSR2, sigusr2_handler)
    # ctrl + c
    signal.signal(signal.SIGINT, sigint_handler)

    # keep process alive
    while True:
        pass

File: packages/dotfiles-utils/src/bin/waifufetch.rs
use clap::Parser;
use dotfiles_utils::NixInfo;
use signal_hook::{
    consts::{SIGINT, SIGUSR2},
    iterator::Signals,
};
use std::process::Command;
use std::{
    io::{self, Write},
    thread,
    time::Duration,
};

fn create_image(nix_info: &NixInfo) -> String {
    let logo = &nix_info.neofetch.logo;
    let hexless = &nix_info.colors;
    let c4 = hexless.get("color4").expect("invalid color");
    let c6 = hexless.get("color6").expect("invalid color");

    let output = format!("/tmp/waifufetch-{}-{}.png", c4, c6);

    let magick_args = [
        logo, // replace color 1
        "-fuzz", "10%", "-fill", c4, "-opaque", "#5278c3", // replace color 2
        "-fuzz", "10%", "-fill", c6, "-opaque", "#7fbae4", &output,
    ];

    Command::new("magick")
        .args(magick_args)
        .status()
        .expect("failed to execute magick");

    output
}

fn waifufetch(nix_info: &NixInfo) {
    let img = create_image(nix_info);
    let neofetch_config = &nix_info.neofetch.conf;

    // assume kitty by default
    let backend = if std::env::var("TERM").unwrap_or("xterm-kitty".to_string()) == "xterm-kitty" {
        "--kitty"
    } else {
        "--sixel"
    };

    let neofetch_args = [backend, &img, "--config", neofetch_config];

    Command::new("neofetch")
        .args(neofetch_args)
        .status()
        .expect("failed to execute neofetch");
}

#[derive(Parser, Debug)]
#[command(name = "waifufetch", about = "Neofetch, but more waifu")]
struct Args {
    #[arg(long, action, help = "prints path to generated image")]
    image: bool,
}

fn main() {
    let args = Args::parse();

    let nix_info = NixInfo::from_cache();

    if args.image {
        println!("{}", create_image(&nix_info));
        std::process::exit(0)
    }

    // initial display of waifufetch
    waifufetch(&nix_info);

    // hide terminal cursor
    print!("\x1B[?25l");
    io::stdout().flush().expect("Failed to flush stdout");

    // handle SIGUSR2 to update colors
    // https://rust-cli.github.io/book/in-depth/signals.html#handling-other-types-of-signals
    let mut signals = Signals::new([SIGINT, SIGUSR2]).unwrap();

    thread::spawn(move || {
        for sig in signals.forever() {
            match sig {
                SIGINT => {
                    // restore terminal cursor
                    print!("\x1B[?25h");
                    io::stdout().flush().expect("Failed to flush stdout");
                    std::process::exit(0);
                }
                SIGUSR2 => {
                    println!("received SIGUSR2");

                    let nix_info = NixInfo::from_cache();
                    waifufetch(&nix_info);
                }
                _ => unreachable!(),
            }
        }
    });

    loop {
        thread::sleep(Duration::from_millis(200));
    }
}
